control/discovery.py

```diff
@@ -0,0 +1,1068 @@
+#
+#  Copyright (c) 2021 International Business Machines
+#  All rights reserved.
+#
+#  SPDX-License-Identifier: LGPL-3.0-or-later
+#
+#  Authors: congmin.yin@intel.com
+#
+
+import argparse
+import grpc
+import json
+import logging
+from .config import GatewayConfig
+from .state import GatewayState, LocalGatewayState, OmapGatewayState, GatewayStateHandler
+
+import rados
+from typing import Dict, Optional
+
+import socket
+import threading
+import time
+import enum
+import uuid
+import struct
+import selectors
+from dataclasses import dataclass, field
+from ctypes import Structure, LittleEndianStructure, c_bool, c_ubyte, c_uint8, c_uint16, c_uint32, c_uint64, c_float
+
+# NVMe tcp pdu type
+class NVME_TCP_PDU(enum.IntFlag):
+    ICREQ = 0x0
+    ICRESP = 0x1
+    H2C_TERM = 0x2
+    C2H_TERM = 0x3
+    CMD = 0x4
+    RSP = 0x5
+    H2C_DATA = 0x6
+    C2H_DATA = 0x7
+    TCP_R2T = 0x9
+
+# NVMe tcp opcode
+class NVME_TCP_OPC(enum.IntFlag):
+    DELETE_SQ = 0x0
+    CREATE_SQ = 0x1
+    GET_LOG_PAGE = 0x2
+    DELETE_CQ = 0x4
+    CREATE_CQ = 0x5
+    IDENTIFY = 0x6
+    ABORT = 0x8
+    SET_FEATURES = 0x9
+    GET_FEATURES = 0xa
+    ASYNC_EVE_REQ = 0xc
+    NS_MGMT = 0xd
+    FW_COMMIT = 0x10
+    FW_IMG_DOWNLOAD = 0x11
+    NS_ATTACH = 0x15
+    KEEP_ALIVE = 0x18
+    FABRIC_TYPE = 0x7F
+
+# NVMe tcp fabric command type
+class NVME_TCP_FCTYPE(enum.IntFlag):
+    PROP_SET = 0x0
+    CONNECT = 0x1
+    PROP_GET = 0x4
+    AUTH_SEND = 0x5
+    AUTH_RECV = 0x6
+    DISCONNECT = 0x8
+
+# NVMe controller register space offsets
+class NVME_CTL(enum.IntFlag):
+    CAPABILITIES = 0x0
+    VERSION = 0x08
+    CONFIGURATION = 0x14
+    STATUS = 0x1c
+
+
+# NVM subsystem types
+class NVMF_SUBTYPE(enum.IntFlag):
+    # Discovery type for NVM subsystem
+    DISCOVERY = 0x1
+    # NVMe type for NVM subsystem
+    NVME = 0x2
+
+# NVMe over Fabrics transport types
+class TRANSPORT_TYPES(enum.IntFlag):
+    RDMA = 0x1
+    FC = 0x2
+    TCP = 0x3
+    INTRA_HOST = 0xfe
+
+# Address family types
+class ADRFAM_TYPES(enum.IntFlag):
+    ipv4 = 0x1
+    ipv6 = 0x2
+    ib = 0x3
+    fc = 0x4
+    intra_host = 0xfe
+
+# Transport requirement, secure channel requirements
+# Connections shall be made over a fabric secure channel
+class NVMF_TREQ_SECURE_CHANNEL(enum.IntFlag):
+    NOT_SPECIFIED = 0x0
+    REQUIRED = 0x1
+    NOT_REQUIRED = 0x2
+
+# maximum number of connections
+MAX_CONNECTION = 10240
+
+# NVMe tcp package length, refer: MTU = 1500 bytes
+NVME_TCP_PDU_UNIT = 1024
+
+# Max SQ head pointer
+SQ_HEAD_MAX = 128
+
+@dataclass
+class Connection(Structure):
+    """Data used multiple times in each connection."""
```

Comment:
I resubmitted the changes and removed `ctypes` from the class `connection`. To avoid sudden requests to access empty `str` objects, I added some judgments. If it is not the expected length, no response will be replied.